README.md
# Scan Splitter
[![pre-commit.ci status](https://results.pre-commit.ci/badge/github/sco1/scan-splitter/main.svg)](https://results.pre-commit.ci/latest/github/sco1/scan-splitter/main)
[![lint-and-test](https://github.com/sco1/scan-splitter/actions/workflows/lint_test.yml/badge.svg?branch=main)](https://github.com/sco1/scan-splitter/actions/workflows/lint_test.yml)
[![Code style: black](https://img.shields.io/badge/code%20style-black-black)](https://github.com/psf/black)
[![Open in Visual Studio Code](https://open.vscode.dev/badges/open-in-vscode.svg)](https://open.vscode.dev/sco1/scan-splitter)

Split composite scan files into separate landmark & measurement files.

## Installation
### Standalone Usage
For users on Windows, self-contained executables are provided for each release.

They can be downloaded from [the project's Releases page](https://github.com/sco1/scan-splitter/releases).

These executables are built from source using [Nuitka](https://github.com/Nuitka/):
```bash
$ python -m nuitka ./src/ui.py --onefile -o scansplitter.exe --plugin-enable=tk-inter --windows-company-name=sco1 --windows-file-version=1.2.1
```

### Via Python
This project utilizes [`poetry`](https://python-poetry.org/) for dependency & environment management. Clone or download this repository to your local machine and create a new environment:

```bash
$ cd <project_dir>
$ poetry install
```

Though it's recommended to utilize `poetry`, the cloned project may also be installed via `pip`:

```bash
$ cd <project_dir>
$ pip install .
```

**NOTE:** Because `pyproject.toml` is present, Poetry must be used in order to perform an editable install, as pip currently does not support editable installs from `pyproject.toml`.

## Usage
Once installed, the `scansplitter` CLI can be invoked directly from the command line:
```bash
$ scansplitter <inputs go here>
```

The `scansplitter` CLI can also be invoked from the root of this repository using Python:
```bash
$ python ./src/ui.py <inputs go here>
```

If using the self-contained executable, the invocation is similar:
```bash
$ ./scansplitter.exe <inputs go here>
```

### `scansplitter --help`
Show the tool's help message & exit.

### `scansplitter`
The bare invocation will execute a streamlined pipeline, prompting the user to select a directory to process & then executing the `batch` and `aggregate` commands with their default values.

For Windows users, this should also be the behavior experienced when double clicking on the self-contained executable.

### `scansplitter single`
Slice the provided scan file and output to CSV.

Inline help may also be viewed using `$ scansplitter single --help`

#### Input Parameters
| Parameter         | Description                          | Type    | Default    |
|-------------------|--------------------------------------|---------|------------|
| `--scan-filepath` | Path to composite scan file to split | Path    | GUI Prompt |

#### Examples
```bash
$ scansplitter single --scan-filepath "./sample_data/067 2021-05-18_06-30-20_composite.txt"
Processing '067 2021-05-18_06-30-20_composite' ... Done!
```

### `scansplitter batch`
Batch process all scans in the specified directory and output to CSVs.

Inline help may also be viewed using `$ scansplitter batch --help`

#### Input Parameters
| Parameter                  | Description                                                       | Type   | Default             |
|----------------------------|-------------------------------------------------------------------|--------|---------------------|
| `--scan-dir`               | Path to directory of composite scan files to split                | Path   | GUI Prompt          |
| `--pattern`                | Glob pattern to use for selecting scan files to split<sup>1</sup> | String | `"*_composite.txt"` |
| `--recurse / --no-recurse` | Recurse through child directories & process all scan files        | Bool   | `False`             |

1. **NOTE:** This scan pattern is assumed to be case-sensitive

#### Examples
```bash
$ scansplitter batch --scan-dir ./sample_data/
Processing '067 2021-05-18_06-30-20_composite' ... Done!
Processing '068 2021-05-18_06-47-57_composite' ... Done!
Processing '069 2021-05-18_07-04-46_composite' ... Done!
```

```bash
$ scansplitter batch --recurse --scan-dir .
Processing '067 2021-05-18_06-30-20_composite' ... Done!
Processing '068 2021-05-18_06-47-57_composite' ... Done!
Processing '069 2021-05-18_07-04-46_composite' ... Done!
```

### `scansplitter aggregate`
Aggregate a directory of split anthro measurement files into a single CSV.

Inline help may also be viewed using `$ scansplitter aggregate --help`

#### Input Parameters
| Parameter                  | Description                                                                         | Type   | Default                    |
|----------------------------|-------------------------------------------------------------------------------------|--------|----------------------------|
| `--anthro-dir`             | Path to directory of anthro files to aggregate                                      | Path   | GUI Prompt                 |
| `--new_names`              | Optional path to a text file for replacement of antho measurement names<sup>1</sup> | Path   | `None`                     |
| `--location_fill`          | Optional fill value for measurement site if missing from filename                   | String | `""`                       |
| `--pattern`                | Glob pattern to use for selecting anthro files to aggregate<sup>2</sup>             | String | `"*_composite.anthro.csv"` |
| `--recurse / --no-recurse` | Recurse through child directories & process all scan files                          | Bool   | `False`                    |

1. **NOTE:** Quantity and order of replacement row names is assumed to match all scans being aggregated. Only quantity is checked before processing.
2. **NOTE:** This scan pattern is assumed to be case-sensitive

#### Examples
```bash
$ scansplitter aggregate
Found 84 anthro measurement files to aggregate.
Using measurement names from: '001 2021-03-31_18-20-36_composite.anthro.csv'
Consolidated measurements file written to: '<data path>/consolidated_anthro.CSV'
```

```bash
$ scansplitter aggregate --location-fill TBS
Found 84 anthro measurement files to aggregate.
Using measurement names from: '001 2021-03-31_18-20-36_composite.anthro.csv'
Consolidated measurements file written to: '<data path>/consolidated_anthro.CSV'
```

```bash
$ scansplitter aggregate --new-row-names "./updated_row_names.txt"
Found 84 anthro measurement files to aggregate.
Using replacement measurement names.
Consolidated measurements file written to: '<data path>/consolidated_anthro.CSV'
```


CHANGELOG.md
# Changelog
Versions follow [Semantic Versioning](https://semver.org/spec/v2.0.0.html) (`<major>`.`<minor>`.`<patch>`)

## [v1.2.1]
### Fixed
* Fixed handling of repeated composite scan files for a given subject (e.g. `123-2` or `123 (2)`).

## [v1.2.0]
### Changed
* The bare `scansplitter` invocation will now execute a streamlined pipeline; The user will be prompted to select a directory to run through the `batch` and `aggregate` pipelines with their respective default values.

## [v1.1.0]
### Added
* Add the `scansplitter aggregate` data aggregation pipeline to aggregate a directory of anthro measurements into a single CSV.

### Changed
* Consolidated scan files are now split into CSVs rather than preserving the formatting of the source file.
* Headers are now written into the measurements & landmarks CSVs when split out of the consolidated scan file.

## [v1.0.0]
Initial release


tests/test_io.py
from pathlib import Path
from textwrap import dedent

import pytest
from src import io


LINE_COUNTER_TEST_CASES = [
    (
        dedent(
            """\
            Measurement Name,Measurement
            Actual Weight,123.45
            Abdomen Circum Tape Measure,3.14
            """
        ),
        3,
    ),
    (
        dedent(
            """\
            Measurement Name,Measurement
            Actual Weight,123.45
            Abdomen Circum Tape Measure,3.14

            """
        ),
        3,
    ),
]


@pytest.mark.parametrize(("raw_src", "truth_count"), LINE_COUNTER_TEST_CASES)
def test_nonempty_line_counter(raw_src: str, truth_count: tuple[str, str]) -> None:  # noqa: D103
    assert io._nonempty_line_count(raw_src) == truth_count


BASE_PATH = Path()
AGGREGATE_HEADER_PREFIX = "Measurement Name"
AGGREGATE_HEADER_TEST_CASES = [
    (
        [
            BASE_PATH / "001 2021-03-31_18-20-36_composite.anthro.csv",
            BASE_PATH / "002 2021-03-31_18-14-36_composite.anthro.csv",
            BASE_PATH / "003 2021-03-31_18-40-49_composite.anthro.csv",
        ],
        "",
        f"{AGGREGATE_HEADER_PREFIX},001,002,003",
    ),
    (
        [
            BASE_PATH / "001 2021-03-31_18-20-36_composite.anthro.csv",
            BASE_PATH / "002 2021-03-31_18-14-36_composite.anthro.csv",
            BASE_PATH / "003 2021-03-31_18-40-49_composite.anthro.csv",
        ],
        "FOO",
        f"{AGGREGATE_HEADER_PREFIX},FOO001,FOO002,FOO003",
    ),
    (
        [
            BASE_PATH / "001 2021-03-31_18-20-36_composite.anthro.csv",
            BASE_PATH / "FOO002 2021-03-31_18-14-36_composite.anthro.csv",
            BASE_PATH / "003 2021-03-31_18-40-49_composite.anthro.csv",
        ],
        "",
        f"{AGGREGATE_HEADER_PREFIX},001,FOO002,003",
    ),
    (
        [
            BASE_PATH / "001 2021-03-31_18-20-36_composite.anthro.csv",
            BASE_PATH / "FOO002 2021-03-31_18-14-36_composite.anthro.csv",
            BASE_PATH / "003 2021-03-31_18-40-49_composite.anthro.csv",
        ],
        "BAR",
        f"{AGGREGATE_HEADER_PREFIX},BAR001,FOO002,BAR003",
    ),
]


@pytest.mark.parametrize(("files", "location_fill", "truth_header"), AGGREGATE_HEADER_TEST_CASES)
def test_aggregate_header_builder(  # noqa: D103
    files: list[Path], location_fill: str, truth_header: str
) -> None:
    assert io._build_aggregate_header(files, AGGREGATE_HEADER_PREFIX, location_fill) == truth_header


MERGER_DUMMY_FILES = [
    "some,header\nmeasurement a,11\nmeasurement b,12",
    "some,header\nmeasurement a,21\nmeasurement b,22",
    "some,header\nmeasurement a,31\nmeasurement b,32",
]
MERGED_ROW_NAMES = ["some header", "measurement a", "measurement b"]
TRUTH_MERGED = [
    "measurement a,11,21,31",
    "measurement b,12,22,32",
]


def test_measurement_merging(tmp_path: Path) -> None:  # noqa: D103
    files = []
    for idx, contents in enumerate(MERGER_DUMMY_FILES, start=1):
        filepath = tmp_path / f"file_{idx:02}.CSV"
        filepath.write_text(contents)
        files.append(filepath)

    merged_measurements = io._merge_measurements(files, MERGED_ROW_NAMES)

    assert merged_measurements == TRUTH_MERGED


tests/test_ui.py
from pytest_mock import MockerFixture
from src import io, ui
from typer.testing import CliRunner

RUNNER = CliRunner()


def test_single_nofile_prompts(mocker: MockerFixture) -> None:
    mocker.patch.object(ui, "_prompt_for_file", autospec=True)
    mocker.patch.object(io, "file_split_pipeline")  # Don't run the pipeline

    result = RUNNER.invoke(ui.scansplitter_cli, ["single"])
    assert result.exit_code == 0
    ui._prompt_for_file.assert_called()


def test_single_file_no_prompt(mocker: MockerFixture) -> None:
    mocker.patch.object(ui, "_prompt_for_file", autospec=True)
    mocker.patch.object(io, "file_split_pipeline")  # Don't run the pipeline

    result = RUNNER.invoke(ui.scansplitter_cli, ["single", "--scan-filepath", "README.md"])
    assert result.exit_code == 0
    ui._prompt_for_file.assert_not_called()


def test_batch_nodir_prompts(mocker: MockerFixture) -> None:
    mocker.patch.object(ui, "_prompt_for_dir", autospec=True)
    mocker.patch.object(io, "batch_split_pipeline")  # Don't run the pipeline

    result = RUNNER.invoke(ui.scansplitter_cli, ["batch"])
    assert result.exit_code == 0
    ui._prompt_for_dir.assert_called()


def test_batch_dir_no_prompt(mocker: MockerFixture) -> None:
    mocker.patch.object(ui, "_prompt_for_dir", autospec=True)
    mocker.patch.object(io, "batch_split_pipeline")  # Don't run the pipeline

    result = RUNNER.invoke(ui.scansplitter_cli, ["batch", "--scan-dir", "."])
    assert result.exit_code == 0
    ui._prompt_for_dir.assert_not_called()


def test_aggregate_nodir_prompts(mocker: MockerFixture) -> None:
    mocker.patch.object(ui, "_prompt_for_dir", autospec=True)
    mocker.patch.object(io, "anthro_measure_aggregation_pipeline")  # Don't run the pipeline

    result = RUNNER.invoke(ui.scansplitter_cli, ["aggregate"])
    assert result.exit_code == 0
    ui._prompt_for_dir.assert_called()


def test_aggregate_dir_no_prompt(mocker: MockerFixture) -> None:
    mocker.patch.object(ui, "_prompt_for_dir", autospec=True)
    mocker.patch.object(io, "anthro_measure_aggregation_pipeline")  # Don't run the pipeline

    result = RUNNER.invoke(ui.scansplitter_cli, ["aggregate", "--anthro-dir", "."])
    assert result.exit_code == 0
    ui._prompt_for_dir.assert_not_called()


def test_bare_invocation_streamlined_pipeline(mocker: MockerFixture) -> None:
    mocker.patch.object(ui, "_prompt_for_dir", autospec=True)
    mocker.patch.object(io, "anthro_measure_aggregation_pipeline")  # Don't run the pipeline
    mocker.patch.object(io, "batch_split_pipeline")  # Don't run the pipeline

    result = RUNNER.invoke(ui.scansplitter_cli)
    assert result.exit_code == 0
    ui._prompt_for_dir.assert_called()
    io.batch_split_pipeline.assert_called()
    io.anthro_measure_aggregation_pipeline.assert_called()


tests/test_parser.py
import itertools
from textwrap import dedent

import click
import pytest
from src import parser


LINE_CLEANER_TEST_CASES = [
    (  # Short-circuit on comment
        "1  *****  Body Fat / Fitness: *****",
        "*****  Body Fat / Fitness: *****",
    ),
    (  # Swap tabs for spaces
        "1  AbdomenBack	5.6	7.8	-9.10",
        "AbdomenBack 5.6 7.8 -9.10",
    ),
    (  # Remove colon
        "1  Actual Weight: 1.2",
        "Actual Weight 1.2",
    ),
    (
        "1  Body Fat (men): 1.2",
        "Body Fat (men) 1.2",
    ),
    (
        "1  Waist at 50%: 1.2",
        "Waist at 50% 1.2",
    ),
    (
        "1  Chest / Bust Circum Tape Measure: 1.2",
        "Chest / Bust Circum Tape Measure 1.2",
    ),
    (
        "1  Bust60DegreesLeft	1.2	1.2	1.2",
        "Bust60DegreesLeft 1.2 1.2 1.2",
    ),
]


@pytest.mark.parametrize(("raw_line", "truth_line"), LINE_CLEANER_TEST_CASES)
def test_line_cleaning(raw_line: str, truth_line: str) -> None:  # noqa: D103
    assert parser._clean_line(raw_line) == truth_line


LINE2CSV_TEST_CASES = [
    (
        "Halter 1.2",
        "Halter,1.2",
    ),
    (
        "halter 1.2",
        "halter,1.2",
    ),
    (
        "Actual Weight 1.2",
        "Actual Weight,1.2",
    ),
    (
        "Chest/Bust Circum Back Left 1.2",
        "Chest/Bust Circum Back Left,1.2",
    ),
    (
        "ActualWeight 1.2",
        "ActualWeight,1.2",
    ),
    (
        "Chest / Bust Circum Tape Measure 1.2",
        "Chest / Bust Circum Tape Measure,1.2",
    ),
    (
        "Body Fat (men) 1.2",
        "Body Fat (men),1.2",
    ),
    (
        "Waist at 50% 1.2",
        "Waist at 50%,1.2",
    ),
    (
        "Bust60DegreesLeft 1.2 1.2 1.2",
        "Bust60DegreesLeft,1.2,1.2,1.2",
    ),
    (
        "AbdomenBack 5.6 7.8 -9.10",
        "AbdomenBack,5.6,7.8,-9.10",
    ),
    (
        "Bust With Drop Back 5.6 7.8 -9.10",
        "Bust With Drop Back,5.6,7.8,-9.10",
    ),
    (
        "Right heel 5.6 7.8 -9.10",
        "Right heel,5.6,7.8,-9.10",
    ),
]


@pytest.mark.parametrize(("raw_line", "truth_csv"), LINE2CSV_TEST_CASES)
def test_line2csv(raw_line: str, truth_csv: str) -> None:  # noqa: D103
    assert parser._line2csv(raw_line) == truth_csv


COMPOSITE_TEST_CASES = [
    (
        dedent(  # Check that headers are discarded; anthro is joined
            """\
            #SizeStream Measurements
            #Stored on Tue May 18 06:49:24 2021
            #SizeStream Core Measurements
            #format - Measurement Valid (1 = valid), Measurement Name, Measurement
            #
            1  Actual Weight: 1.2
            #SizeStream Custom Measurements
            #format - Measurement Valid (1 = valid), Measurement Name, Measurement
            #
            1  Chest: 3.4
            #SizeStream Landmarks
            #format - Landmarks Valid (1 = valid), Landmark Name, Landmark x y z
            #
            1  AbdomenBack	5.6	7.8	-9.10
            """
        ),
        ["Actual Weight,1.2", "Chest,3.4"],
        ["AbdomenBack,5.6,7.8,-9.10"],
    ),
    (
        dedent(  # Check that comments are discarded
            """\
            #SizeStream Measurements
            #Stored on Tue May 18 06:49:24 2021
            #SizeStream Core Measurements
            #format - Measurement Valid (1 = valid), Measurement Name, Measurement
            #
            1  Actual Weight: 1.2
            #SizeStream Custom Measurements
            #format - Measurement Valid (1 = valid), Measurement Name, Measurement
            #
            1  *****  Body Fat / Fitness: *****
            1  Chest: 3.4
            #SizeStream Landmarks
            #format - Landmarks Valid (1 = valid), Landmark Name, Landmark x y z
            #
            1  AbdomenBack	5.6	7.8	-9.10
            """
        ),
        ["Actual Weight,1.2", "Chest,3.4"],
        ["AbdomenBack,5.6,7.8,-9.10"],
    ),
]


@pytest.mark.parametrize(
    ("raw_src", "truth_anthro", "truth_landmark"), COMPOSITE_TEST_CASES, ids=itertools.count()
)
def test_composite_file_parsing(  # noqa: D103
    raw_src: str, truth_anthro: list[str], truth_landmark: list[str]
) -> None:
    anthro, landmark = parser.split_composite_file(raw_src.splitlines())

    assert anthro == truth_anthro
    assert landmark == truth_landmark


SUBJ_ID_TEST_CASES = [
    (
        "1 2021-04-20_18-00-00_composite",
        ("1", ""),
    ),
    (
        "001 2021-04-20_18-00-00_composite",
        ("001", ""),
    ),
    (
        "CPEN1 2021-04-20_18-00-00_composite",
        ("1", "CPEN"),
    ),
    (
        "CPEN001 2021-04-20_18-00-00_composite",
        ("001", "CPEN"),
    ),
    (
        "cpen001 2021-04-20_18-00-00_composite",
        ("001", "cpen"),
    ),
    (
        "A001 2021-04-20_18-00-00_composite",
        ("001", "A"),
    ),
    (
        "A1 2021-04-20_18-00-00_composite",
        ("1", "A"),
    ),
    (
        "a1 2021-04-20_18-00-00_composite",
        ("1", "a"),
    ),
    (
        "1234 (2) 2021-04-20_18-00-00_composite",
        ("1234 (2)", ""),
    ),
    (
        "CPEN1234 (2) 2021-04-20_18-00-00_composite",
        ("1234 (2)", "CPEN"),
    ),
    (
        "1234-2 2021-04-20_18-00-00_composite",
        ("1234-2", ""),
    ),
    (
        "CPEN1234-2 2021-04-20_18-00-00_composite",
        ("1234-2", "CPEN"),
    ),
]


@pytest.mark.parametrize(("raw_line", "truth_id"), SUBJ_ID_TEST_CASES)
def test_subject_id_extraction(raw_line: str, truth_id: tuple[str, str]) -> None:  # noqa: D103
    assert parser.extract_subj_id(raw_line) == truth_id


def test_no_subject_id_raises() -> None:  # noqa: D103
    with pytest.raises(click.ClickException):
        parser.extract_subj_id("2021-04-20_18-00-00_composite")


SAMPLE_DEFAULT_LOCATION = "FOO"
LOCATION_INSERTION_TEST_CASES = [
    (
        "001 2021-04-20_18-00-00_composite",
        ("001", SAMPLE_DEFAULT_LOCATION),
    ),
    (
        "CPEN001 2021-04-20_18-00-00_composite",
        ("001", "CPEN"),
    ),
]


@pytest.mark.parametrize(("raw_line", "truth_id"), LOCATION_INSERTION_TEST_CASES)
def test_default_location_insertion(raw_line: str, truth_id: tuple[str, str]) -> None:  # noqa: D103
    assert parser.extract_subj_id(raw_line, default_location=SAMPLE_DEFAULT_LOCATION) == truth_id


ROW_EXTRACTION_TEST_CASES = [
    (
        dedent(
            """\
            Measurement Name,Measurement
            Actual Weight,123.45
            Abdomen Circum Tape Measure,3.14
            """
        ),
        [
            "Measurement Name",
            "Actual Weight",
            "Abdomen Circum Tape Measure",
        ],
    ),
    (
        dedent(
            """\
            SS_Vars
            Actual_Weight
            Abdomen_Circum_Tape_Measure
            """
        ),
        [
            "SS_Vars",
            "Actual_Weight",
            "Abdomen_Circum_Tape_Measure",
        ],
    ),
]


@pytest.mark.parametrize(("raw_src", "truth_rownames"), ROW_EXTRACTION_TEST_CASES)
def test_row_name_extraction(raw_src: str, truth_rownames: list[str]) -> None:  # noqa: D103
    assert parser.extract_measurement_names(raw_src) == truth_rownames


src/ui.py
import tkinter as tk
from pathlib import Path
from tkinter import filedialog

import click
import typer
from src import io


scansplitter_cli = typer.Typer()


def _prompt_for_file(title: str, start_dir: Path = Path()) -> Path:  # pragma: no cover
    """Open a Tk file selection dialog to prompt the user to select a single file for processing."""
    root = tk.Tk()
    root.withdraw()

    picked = filedialog.askopenfilename(  # type: ignore[no-untyped-call]  # stubs need mypy bump
        title=title,
        initialdir=start_dir,
        multiple=False,
        filetypes=[
            ("Composite Scan Data", "*.txt"),
            ("All Files", "*.*"),
        ],
    )

    if not picked:
        raise click.ClickException("No file selected for parsing, aborting.")

    return Path(picked)


def _prompt_for_dir(start_dir: Path = Path()) -> Path:  # pragma: no cover
    """Open a Tk file selection dialog to prompt the user to select a directory for processing."""
    root = tk.Tk()
    root.withdraw()

    picked = filedialog.askdirectory(  # type: ignore[no-untyped-call]  # stubs need mypy bump
        title="Select directory for batch processing",
        initialdir=start_dir,
    )

    if not picked:
        raise click.ClickException("No directory selected for parsing, aborting.")

    return Path(picked)


@scansplitter_cli.command()
def single(
    scan_filepath: Path = typer.Option(None, exists=True, file_okay=True, dir_okay=False),
) -> None:
    """
    Split the specified scan file into its anthro & landmark components.

    If no file is specified, the user will be prompted to select one.
    """
    if scan_filepath is None:
        scan_filepath = _prompt_for_file(title="Select scan file to slice")

    io.file_split_pipeline(scan_filepath)


@scansplitter_cli.command()
def batch(
    scan_dir: Path = typer.Option(None, exists=True, file_okay=False, dir_okay=True),
    pattern: str = typer.Option("*_composite.txt"),
    recurse: bool = False,
) -> None:
    """
    Batch process all scans in the specified directory.

    If no processing directory is specified, the user will be prompted to select one.

    Recursive processing may be optionally specified (Default: `False`).
    """
    if scan_dir is None:
        scan_dir = _prompt_for_dir()

    io.batch_split_pipeline(scan_dir, pattern=pattern, recurse=recurse)


@scansplitter_cli.command()
def aggregate(
    anthro_dir: Path = typer.Option(None, exists=True, file_okay=False, dir_okay=True),
    new_row_names: Path = typer.Option(None, exists=True, file_okay=True, dir_okay=False),
    location_fill: str = "",
    pattern: str = typer.Option("*_composite.anthro.csv"),
    recurse: bool = False,
) -> None:
    """
    Aggregate a directory of split anthro measurement files into a single CSV.

    If no processing directory is specified, the user will be prompted to select one.

    Recursive processing may be optionally specified (Default: `False`).
    """
    if anthro_dir is None:
        anthro_dir = _prompt_for_dir()

    io.anthro_measure_aggregation_pipeline(
        anthro_dir,
        new_row_names=new_row_names,
        location_fill=location_fill,
        pattern=pattern,
        recurse=recurse,
    )


@scansplitter_cli.callback(invoke_without_command=True)
def main(ctx: typer.Context) -> None:
    """
    Split composite scan file(s) into separate landmark & measurement files.

    Invocation without args prompts the user to select a scan directory, then runs through the
    default batch & aggregation pipelines.
    """
    if not ctx.invoked_subcommand:
        anthro_dir = _prompt_for_dir()
        io.batch_split_pipeline(anthro_dir)
        io.anthro_measure_aggregation_pipeline(anthro_dir)


if __name__ == "__main__":  # pragma: no cover
    scansplitter_cli()


src/io.py
import typing as t
from pathlib import Path

from rich import print as rprint
from src import parser


# Default Headers
ANTHRO_HEADER = ["Measurement Name,Measurement"]
LANDMARK_HEADER = ["Landmark Name,x,y,z"]


def _dump_chunk(filepath: Path, data: list[str], header: t.Optional[list[str]] = None) -> None:
    """
    Write the input header & data line(s) to the provided output filepath.

    NOTE: Any existing file will be overwritten
    """
    with filepath.open("w") as f:
        if header:
            # Headers need a trailing newline since they'll be followed by our data
            f.write("".join(f"{line}\n" for line in header))

        f.write("\n".join(data))


def file_split_pipeline(in_file: Path) -> None:
    """
    Split the provided composite file into CSVs of its anthro & landmark components.

    Split files are written to the same directory as the input file:
        * Anthro components append ".anthro" to the name of the file
        * Landmark components append ".lmk" to the name of the file

    e.g. `./some_scan_composite.txt` becomes:
        `./some_scan_composite.anthro.csv` and `./some_scan_composite.lmk.csv`

    Comments (lines containing `*`) and header lines (lines beginning with `#`) are discarded

    NOTE: Any existing anthro & landmark files will be overwritten
    """
    base_stem = in_file.stem
    anthro_filepath = in_file.with_name(f"{base_stem}.anthro.csv")
    landmark_filepath = in_file.with_name(f"{base_stem}.lmk.csv")

    rprint(f"Processing {base_stem!r} ... ", end="")

    composite_src = in_file.read_text().splitlines()
    anthro, landmark = parser.split_composite_file(composite_src)

    _dump_chunk(anthro_filepath, anthro, ANTHRO_HEADER)
    _dump_chunk(landmark_filepath, landmark, LANDMARK_HEADER)

    rprint("[green]Done!")


def batch_split_pipeline(
    in_dir: Path, pattern: str = "*_composite.txt", recurse: bool = False
) -> None:
    """
    Batch process all files in the specified directory that match the provided glob pattern.

    Recursion can optionally be specified by `recurse`.

    NOTE: If `recurse` is `True`, do not include `**` in `pattern`, this is not guarded against.
    """
    if recurse:
        pattern = f"**/{pattern}"

    n = 0
    for composite_file in in_dir.glob(pattern):
        file_split_pipeline(composite_file)
        n += 1

    rprint(f"Processed {n} files")


def _nonempty_line_count(src: str) -> int:
    """Count the number of non-empty lines present in the provided source string."""
    return sum(1 for line in src.splitlines() if line.strip())


def _has_same_n_rows(data_file: Path, new_row_names: Path) -> bool:
    """Check that the two files provided have the same number of non-empty lines."""
    return _nonempty_line_count(data_file.read_text()) == _nonempty_line_count(
        new_row_names.read_text()
    )


def _build_aggregate_header(files: list[Path], header_prefix: str, location_fill: str) -> str:
    """Generate the header line for the aggregate measurement CSV."""
    col_names = []
    for file in files:
        subj_id, location = parser.extract_subj_id(file.name, location_fill)
        col_names.append(f"{location}{subj_id}")

    return f"{header_prefix},{','.join(col_names)}"


def _merge_measurements(files: list[Path], row_names: list[str]) -> list[str]:
    """
    Merge measurement values from the provided list of anthro measurement files.

    Files are assumed to be of the format output by our scan splitting pipeline. Measurements are
    merged into a list of strings, one measurement per row.

    e.g.:
        some,header
        measurement a,11
        measurement b,12

        some,header
        measurement a,21
        measurement b,22

    Becomes:
        ["measurement a,11,21", "measurement b,12,22"]

    Row names and order are assumed to be consistent across all input files, as well as the input
    list of row names
    """
    # Iterate through all of the anthro measurement files & pull in the entire measurements column
    # for each file & store into a list of lists
    all_measurements = []
    for file in files:
        data_lines = file.read_text().splitlines()[1:]  # skip header line
        values = [line.split(",")[1] for line in data_lines]
        all_measurements.append(values)

    # Since we have a list of columns, we can use zip to join them into a row for each column
    # We can also add the row names (sans header) in with this step
    joined_measurements = [",".join(line) for line in zip(row_names[1:], *all_measurements)]

    return joined_measurements


def anthro_measure_aggregation_pipeline(
    anthro_dir: Path,
    new_row_names: t.Optional[Path] = None,
    location_fill: str = "",
    pattern: str = "*_composite.anthro.csv",
    recurse: bool = False,
) -> None:
    """Aggregate a directory of split anthro measurement files into a single CSV."""
    if recurse:
        pattern = f"**/{pattern}"

    # Listify here so we can run some short-circuit checks on a sample file before launching into
    # the rest of the pipeline
    anthro_files = list(anthro_dir.glob(pattern))
    if not anthro_files:
        rprint(f"No files found in '{anthro_dir}' matching '{pattern}'")
        return
    else:
        rprint(f"Found {len(anthro_files)} anthro measurement files to aggregate.")

    # Get measurement (row) names, either from the first measurement file or from the specified
    # replacement file
    # All measurement files are assumed to contain the same number & order of measurements
    if new_row_names:
        # Do a basic check to see if the replacement file has the same number of rows
        # Both files are assumed to contain one header line
        if not _has_same_n_rows(anthro_files[0], new_row_names):
            rprint(
                f"Length mismatch between anthro files & replacement measurement names, please check your file: '{new_row_names}'"  # noqa: E501
            )
            return
        else:
            rprint("Using replacement measurement names.")
            row_names = parser.extract_measurement_names(new_row_names.read_text())
    else:
        rprint(f"Using measurement names from: '{anthro_files[0].name}'")
        row_names = parser.extract_measurement_names(anthro_files[0].read_text())

    # Build the aggregate header line, which appends all of the subject IDs to the header of the row
    # names
    aggregate_header = _build_aggregate_header(
        anthro_files, header_prefix=row_names[0], location_fill=location_fill
    )
    joined_measurements = _merge_measurements(anthro_files, row_names)

    out_filepath = anthro_dir / "consolidated_anthro.CSV"
    _dump_chunk(out_filepath, joined_measurements, [aggregate_header])
    rprint(f"Consolidated measurements file written to: '{out_filepath}'")
